Remove `String::size`. Use `String::length` instead.

Ivorforce opened the pull request:
`String::size` is error prone and misleading. I am proposing to remove it.
`String::size` is luckily not exposed to GDExtension or GDScript.

This PR is likely to both fix subtle bugs and cause regressions. I recommend merging it soon (early in the 4.5 release cycle).

## Explanation

I was investigating why https://github.com/godotengine/godot/pull/102369 failed, when I realized that empty strings have a dichotomous state:
- `_cowdata` is empty (`_ptr == nullptr`)
- `_cowdata` is allocated and `_size` is 1. The only character is used for the trailing 0 (`NUL` terminator).

This peculiarity does not seem to be widely known. It is currently common to check for `String` emptiness like this:

https://github.com/godotengine/godot/blob/931820d33cbc38328e4014d65dd62c359b5116fc/drivers/gles3/shader_gles3.cpp#L179

This is incorrect because the `String` in question might be empty but allocated, whereupon `size()` would return 1. It should call `if (p_version->uniforms.length())` instead (or better, `if (!p_version->uniforms.is_empty())`).

Additionally, `size` is currently often mistaken for `length` in the codebase. While making this PR, I found many subtle bugs of code that was iterating over the trailing `NUL` character and (potentially) inserting it into strings or other data.

All in all, `size` doesn't really make sense for a `String`. Almost all callers should use `length()` instead. The trailing `NUL` is only relevant for string parsers that use pointers and an implicit trailing `NUL` to terminate (and thus not to anyone using string length semantics).

## Discussion

It would be possible to keep the old `size()` around, renamed to `buffer_size()` to avoid mistakes.
I opted not to do this because I don't see many use-cases for it, but I could be persuaded otherwise.

## Addendum

We should resolve the dichotomy mentioned above. Empty `String`s should not allocate. Without the dichotomy, https://github.com/godotengine/godot/pull/102369 can also be merged for a pretty significant speed boost.
This will unfortunately require a little bit of a compatibility breakage, so I wanted to remove `size()` first.

Mickeon commented:
> Additionally, `size` is currently often mistaken for `length` in the codebase.

Not just in the codebase. See:
- https://github.com/godotengine/godot-proposals/issues/7869

Ivorforce committed:
Remove `String::size`. Use `String::length` instead.
`String::size` was error prone because empty strings may (currently) be of both length 0 and length 1. Additionally, it was often being mistaken for `length`.

Ivorforce commented:
I've implemented a solution that just renames `size()` to `buffer_size` and continues to use that for cases where it is unclear to me how the surrounding code makes use of it.
It ended up being a lot less risky and still solves the problem. We can probably get rid of more `buffer_size` uses passively as the code is being churned.

I'll mark this PR as draft and create the other PR once it compiles on all platforms.
I think it's the better solution to the problem at hand.
